### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by UladzimirTrehubenka**

### CODE REVIEW

File: contrib/coredns/policy/policy.go [hunk 1]
```diff
@@ -272,139 +270,170 @@ func (p *PolicyPlugin) decodeDebugMsg(r *dns.Msg) (*dns.Msg, bool) {
...
 				if err != nil {
-					return p.retRcode(w, r, dns.RcodeServerFailure, err)
+					status = dns.RcodeServerFailure
+					goto Exit
 				}
+			} else {
+				resolve = resolveNo
 			}
+		} else {
+			resolve = resolveFailed
```
Comment (rdrozhdzh) [inline contrib/coredns/policy/policy.go]: don't we want to change status to dns.RcodeServerFailure here?

File: contrib/coredns/policy/policy.go [hunk 2]
```diff
@@ -272,139 +270,170 @@ func (p *PolicyPlugin) decodeDebugMsg(r *dns.Msg) (*dns.Msg, bool) {
...
+	default:
+		status = dns.RcodeServerFailure
+		err = errInvalidAction
 	}
 
-	return p.retRcode(w, r, dns.RcodeServerFailure, errInvalidAction)
+Exit:
+	r.Rcode = status
+	r.Response = true
+	if p.TapIO != nil {
```
Comment (rdrozhdzh) [inline contrib/coredns/policy/policy.go]: && status != dns.RcodeServerFailure
don't send dnstap msg if server failed

File: contrib/coredns/policy/policy.go [hunk 3]
```diff
@@ -272,139 +270,170 @@ func (p *PolicyPlugin) decodeDebugMsg(r *dns.Msg) (*dns.Msg, bool) {
...
+		status, err := plugin.NextOrFailure(p.Name(), p.Next, ctx, responseWriter, r)
+		r.Question[0].Name = origName
 		if err != nil {
-			return p.retRcode(w, r, dns.RcodeServerFailure, err)
+			return dns.RcodeServerFailure, err
 		}
-
-		a.Answer = append(a.Answer, responseWriter.Msg.Answer...)
-		a.SetRcode(r, status)
+		r.Answer = []dns.RR{rr}
```
Comment (rdrozhdzh) [inline contrib/coredns/policy/policy.go]: common code for "if" and "else", can be moved before "if" statement

Comment (UladzimirTrehubenka) [reply contrib/coredns/policy/policy.go]: r.answer affects on NextOrFailure() so we cannot move this line up

File: contrib/coredns/policy/attrholder.go [hunk 4]
```diff
@@ -84,18 +73,41 @@ func (ah *attrHolder) request() []*pdp.Attribute {
 }
 
 func (ah *attrHolder) addResponse(r *pdp.Response) {
-	a := r.Obligation
+	if r == nil {
```
Comment (vasili-v) [inline contrib/coredns/policy/attrholder.go]: Why do you need this? It anyway will panic later

File: pep/benchmark_test.go [hunk 5]
```diff
@@ -1,216 +1,19 @@
...
-	"strings"
-	"sync"
 	"testing"
 	"time"
 
 	pb "github.com/infobloxopen/themis/pdp-service"
 )
 
 const (
-	oneStageBenchmarkPolicySet = `# Policy set for benchmark
```
Comment (vasili-v) [inline pep/benchmark_test.go]: Don't remove the cases

File: pep/benchmark_test.go [hunk 6]
```diff
@@ -779,336 +523,35 @@ func makeTempFile(s, prefix string) (string, error) {
...
-func (p *pipe) cleanup() []string {
-	if p.w != nil {
-		p.w.Close()
-		p.w = nil
-	}
-
-	if p.r != nil {
-		p.r.Close()
-		p.r = nil
-		p.Wait()
```
Comment (vasili-v) [inline pep/benchmark_test.go]: The only fix needed here is just to move p.Wait to previous if block

File: pep/benchmark_test.go [hunk 7]
```diff
@@ -719,42 +470,35 @@ func BenchmarkThreeStagePolicySetRaw(b *testing.B) {
...
 
 	tmpJCon, err := makeTempFile(benchmarkContent, "content")
 	if err != nil {
 		os.Remove(tmpYAST)
-		b.Fatalf("can't create content file: %s", err)
+		return "", "", nil, fmt.Errorf("cannot create content file: %s", err)
 	}
 
-	pdp, err := newProc("pdpserver", "-p", tmpYAST, "-j", tmpJCon)
+	process, err := startProcess("../build/pdpserver", "-p", tmpYAST, "-j", tmpJCon)
```
Comment (vasili-v) [inline pep/benchmark_test.go]: The relative path implies particular directory to run benchmark. I'd change to $GOPATH/src/github.com/infobloxopen/themis/build/pdpserver

**Pull request merged**